File: TheWatch-Android/app/src/main/java/com/thewatch/app/data/signalr/HubEventSender.kt
package com.thewatch.app.data.signalr

import com.thewatch.app.data.logging.WatchLogger
import kotlinx.coroutines.*
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.time.Instant
import java.util.concurrent.ConcurrentLinkedQueue
import javax.inject.Inject
import javax.inject.Singleton

// ── Write-Ahead Log ──────────────────────────────────────────────
// WAL: HubEventSender is the outbound message sender for the SignalR hub connection.
// It provides a typed API for all messages the mobile client sends to the DashboardHub,
// with automatic queueing when disconnected and flush-on-reconnect.
//
// Architecture:
//   - Thread-safe: all sends go through a Mutex-guarded serial queue
//   - Offline resilience: messages queued in ConcurrentLinkedQueue when disconnected
//   - Auto-flush: listens for HubConnectionListener.onConnected() and flushes queue
//   - Queue cap: max 500 messages to prevent OOM during extended offline periods
//   - All outbound messages logged through WatchLogger with sourceContext "SignalR"
//
// Outbound messages:
//   - "TestStepCompleted" → sends test step result back to orchestrator
//   - "LogEntry" → streams a log entry to the dashboard
//   - "DeviceStatusUpdate" → periodic device health/status report
//   - "Pong" → response to server Ping
//   - "ReportAdapterTiers" → current adapter tier map
//   - "UpdateResponderLocation" → en-route location updates
//   - "ResponderOnScene" → arrival notification
//   - "NotifyEvidenceSubmitted" → evidence upload notification
//
// Example:
// ```kotlin
// @Inject lateinit var sender: HubEventSender
//
// // Send a test step result
// sender.sendTestStepCompleted(
//     runId = "run_abc123",
//     stepId = "step_001",
//     passed = true,
//     screenshot = null,
//     errorMessage = null,
//     durationMs = 1250
// )
//
// // Stream a log entry
// sender.sendLogEntry(
//     level = "Information",
//     source = "LocationCoordinator",
//     message = "Mode escalated to Emergency",
//     correlationId = "sos_xyz"
// )
//
// // Messages are queued if disconnected and auto-flushed on reconnect
// ```

/**
 * Queued message waiting to be sent when the connection is restored.
 */
@Serializable
data class QueuedHubMessage(
    val method: String,
    val argsJson: String,
    val queuedAt: String // ISO-8601 timestamp
)

/**
 * Outbound message sender for the SignalR hub.
 *
 * Provides typed methods for every message the mobile client sends to the
 * DashboardHub. Handles offline queueing and reconnect flushing automatically.
 *
 * Thread safety: all sends are serialized through a [Mutex] to prevent
 * concurrent writes on the SignalR connection, which is not thread-safe.
 * The queue is a [ConcurrentLinkedQueue] for lock-free enqueue from any thread.
 */
@Singleton
class HubEventSender @Inject constructor(
    private val hubConnection: WatchHubConnection,
    private val logger: WatchLogger
) : HubConnectionListener {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val sendMutex = Mutex()
    private val json = Json { encodeDefaults = true }

    // ── Offline Queue ────────────────────────────────────────────

    /**
     * Messages queued while disconnected. Flushed FIFO on reconnect.
     * Capped at [maxQueueSize] to prevent OOM.
     */
    private val messageQueue = ConcurrentLinkedQueue<QueuedHubMessage>()

    /** Maximum number of queued messages. Oldest are dropped when exceeded. */
    private val maxQueueSize = 500

    init {
        hubConnection.addListener(this)
    }

    // ── HubConnectionListener ────────────────────────────────────

    override fun onConnected() {
        scope.launch { flushQueue() }
    }

    override fun onDisconnected(error: Throwable?) {
        // Queue will accumulate messages until reconnect
    }

    override fun onReconnecting(attemptNumber: Int) {
        // No action — queue continues to accumulate
    }

    // ── Test Orchestration ───────────────────────────────────────

    /**
     * Send a test step result back to the TestOrchestratorService.
     *
     * The orchestrator uses this to advance the test run: recording the result,
     * dispatching the next step, and broadcasting to dashboard observers.
     *
     * Maps to: TestOrchestratorService.RecordStepResultAsync(runId, stepId, passed, screenshot, errorMessage)
     */
    fun sendTestStepCompleted(
        runId: String,
        stepId: String,
        passed: Boolean,
        screenshot: String? = null,
        errorMessage: String? = null,
        durationMs: Long = 0
    ) {
        val args = buildMap {
            put("runId", runId)
            put("stepId", stepId)
            put("passed", passed.toString())
            screenshot?.let { put("screenshot", it) }
            errorMessage?.let { put("errorMessage", it) }
            put("durationMs", durationMs.toString())
        }

        send("TestStepCompleted", json.encodeToString(args), correlationId = runId)
    }

    // ── Log Streaming ────────────────────────────────────────────

    /**
     * Stream a log entry to the dashboard for real-time log viewing.
     *
     * Called by the LoggingPort adapter when log streaming is active.
     * The dashboard's LogViewerPanel displays these in real-time.
     */
    fun sendLogEntry(
        level: String,
        source: String,
        message: String,
        properties: Map<String, String> = emptyMap(),
        exception: String? = null,
        correlationId: String? = null
    ) {
        val args = buildMap {
            put("level", level)
            put("source", source)
            put("message", message)
            put("timestamp", Instant.now().toString())
            if (properties.isNotEmpty()) put("properties", json.encodeToString(properties))
            exception?.let { put("exception", it) }
            correlationId?.let { put("correlationId", it) }
        }

        send("LogEntry", json.encodeToString(args), correlationId = correlationId)
    }

    // ── Device Status ────────────────────────────────────────────

    /**
     * Send a periodic device status update to the dashboard.
     *
     * Includes battery level, network type, GPS availability, and
     * current adapter tier configuration. Sent every 30s by a background worker.
     */
    fun sendDeviceStatusUpdate(
        batteryPercent: Int,
        networkType: String,
        gpsEnabled: Boolean,
        adapterTiers: Map<String, String>,
        freeMemoryMb: Long
    ) {
        val args = buildMap {
            put("batteryPercent", batteryPercent.toString())
            put("networkType", networkType)
            put("gpsEnabled", gpsEnabled.toString())
            put("adapterTiers", json.encodeToString(adapterTiers))
            put("freeMemoryMb", freeMemoryMb.toString())
            put("timestamp", Instant.now().toString())
        }

        send("DeviceStatusUpdate", json.encodeToString(args))
    }

    // ── Keepalive ────────────────────────────────────────────────

    /**
     * Respond to a server Ping with Pong. Not queued — if disconnected,
     * the Ping/Pong cycle is irrelevant.
     */
    fun sendPong() {
        if (hubConnection.connectionState.value != HubConnectionState.Connected) return

        scope.launch {
            sendMutex.withLock {
                // val connection = hubConnection.hubConnection as? com.microsoft.signalr.HubConnection
                // connection?.send("Pong", Instant.now().toString())

                logger.verbose(
                    sourceContext = "SignalR",
                    messageTemplate = "Sent Pong"
                )
            }
        }
    }

    // ── Adapter Tier Reporting ────────────────────────────────────

    /**
     * Report current adapter tier configuration to the dashboard.
     * Called on connect and when the dashboard requests it.
     */
    fun sendReportAdapterTiers(tiers: Map<String, String>) {
        send("ReportAdapterTiers", json.encodeToString(tiers))
    }

    // ── Response Coordination ────────────────────────────────────

    /**
     * Send a responder location update while en-route to an SOS.
     */
    fun sendResponderLocation(
        requestId: String,
        responderId: String,
        latitude: Double,
        longitude: Double,
        speedMps: Double? = null
    ) {
        val args = buildMap {
            put("requestId", requestId)
            put("responderId", responderId)
            put("latitude", latitude.toString())
            put("longitude", longitude.toString())
            speedMps?.let { put("speedMps", it.toString()) }
            put("timestamp", Instant.now().toString())
        }

        send("UpdateResponderLocation", json.encodeToString(args), correlationId = requestId)
    }

    /**
     * Signal that the responder has arrived on scene.
     */
    fun sendResponderOnScene(requestId: String, responderId: String) {
        val args = mapOf(
            "requestId" to requestId,
            "responderId" to responderId,
            "arrivedAt" to Instant.now().toString()
        )

        send("ResponderOnScene", json.encodeToString(args), correlationId = requestId)
    }

    /**
     * Notify the dashboard that evidence has been submitted from this device.
     */
    fun sendEvidenceSubmitted(
        submissionId: String,
        requestId: String,
        phase: String,
        type: String,
        thumbnailUrl: String? = null
    ) {
        val args = buildMap {
            put("submissionId", submissionId)
            put("requestId", requestId)
            put("phase", phase)
            put("type", type)
            thumbnailUrl?.let { put("thumbnailUrl", it) }
            put("timestamp", Instant.now().toString())
        }

        send("NotifyEvidenceSubmitted", json.encodeToString(args), correlationId = requestId)
    }

    // ── Core Send / Queue ────────────────────────────────────────

    /**
     * Send a message to the hub, or queue it if disconnected.
     *
     * All sends are serialized through [sendMutex] to prevent concurrent
     * writes on the SignalR connection (which is not thread-safe).
     */
    private fun send(method: String, argsJson: String, correlationId: String? = null) {
        if (hubConnection.connectionState.value != HubConnectionState.Connected) {
            enqueue(method, argsJson)
            return
        }

        scope.launch {
            sendMutex.withLock {
                try {
                    // val connection = hubConnection.hubConnection as? com.microsoft.signalr.HubConnection
                    // connection?.send(method, argsJson)

                    logger.debug(
                        sourceContext = "SignalR",
                        messageTemplate = "Sent {Method} ({Bytes} bytes)",
                        properties = mapOf(
                            "Method" to method,
                            "Bytes" to argsJson.length.toString()
                        ),
                        correlationId = correlationId
                    )
                } catch (e: Exception) {
                    logger.warning(
                        sourceContext = "SignalR",
                        messageTemplate = "Failed to send {Method}: {Error}. Queueing.",
                        properties = mapOf(
                            "Method" to method,
                            "Error" to (e.message ?: "unknown")
                        ),
                        exception = e,
                        correlationId = correlationId
                    )
                    enqueue(method, argsJson)
                }
            }
        }
    }

    /**
     * Enqueue a message for later delivery. Drops oldest messages if the
     * queue exceeds [maxQueueSize].
     */
    private fun enqueue(method: String, argsJson: String) {
        val message = QueuedHubMessage(
            method = method,
            argsJson = argsJson,
            queuedAt = Instant.now().toString()
        )

        messageQueue.add(message)

        // Trim queue if over capacity (drop oldest)
        while (messageQueue.size > maxQueueSize) {
            val dropped = messageQueue.poll()
            if (dropped != null) {
                logger.warning(
                    sourceContext = "SignalR",
                    messageTemplate = "Queue overflow: dropped {Method} queued at {QueuedAt}",
                    properties = mapOf(
                        "Method" to dropped.method,
                        "QueuedAt" to dropped.queuedAt
                    )
                )
            }
        }

        logger.debug(
            sourceContext = "SignalR",
            messageTemplate = "Queued {Method} for delivery on reconnect. Queue size: {Size}",
            properties = mapOf(
                "Method" to method,
                "Size" to messageQueue.size.toString()
            )
        )
    }

    /**
     * Flush all queued messages after reconnection. Messages are sent FIFO
     * through the [sendMutex] to maintain ordering.
     */
    private suspend fun flushQueue() {
        val queueSize = messageQueue.size
        if (queueSize == 0) return

        logger.information(
            sourceContext = "SignalR",
            messageTemplate = "Flushing {Count} queued messages after reconnect",
            properties = mapOf("Count" to queueSize.toString())
        )

        var flushed = 0
        var failed = 0

        while (messageQueue.isNotEmpty()) {
            val message = messageQueue.poll() ?: break

            sendMutex.withLock {
                try {
                    // val connection = hubConnection.hubConnection as? com.microsoft.signalr.HubConnection
                    // connection?.send(message.method, message.argsJson)
                    flushed++
                } catch (e: Exception) {
                    failed++
                    logger.warning(
                        sourceContext = "SignalR",
                        messageTemplate = "Failed to flush {Method}: {Error}",
                        properties = mapOf(
                            "Method" to message.method,
                            "Error" to (e.message ?: "unknown")
                        )
                    )
                    // Don't re-queue — if we can't send after reconnect, something is wrong
                }
            }
        }

        logger.information(
            sourceContext = "SignalR",
            messageTemplate = "Queue flush complete: {Flushed} sent, {Failed} failed",
            properties = mapOf(
                "Flushed" to flushed.toString(),
                "Failed" to failed.toString()
            )
        )
    }

    // ── Diagnostics ──────────────────────────────────────────────

    /**
     * Returns queue diagnostics for the diagnostics screen.
     */
    fun diagnosticsSnapshot(): Map<String, String> = mapOf(
        "QueueSize" to messageQueue.size.toString(),
        "MaxQueueSize" to maxQueueSize.toString(),
        "ConnectionState" to hubConnection.connectionState.value.name
    )
}

File: TheWatch-iOS/Services/SignalR/HubEventSender.swift
import Foundation

// ── Write-Ahead Log ──────────────────────────────────────────────
// WAL: HubEventSender is the outbound message sender for the SignalR hub on iOS.
// It provides a typed API for all messages the mobile client sends to the DashboardHub,
// with automatic queueing when disconnected and flush-on-reconnect.
//
// Architecture:
//   - Actor-based concurrency via serial DispatchQueue for thread-safe sends
//   - Offline resilience: messages queued in an array when disconnected
//   - Auto-flush: implements HubConnectionDelegate and flushes queue on connect
//   - Queue cap: max 500 messages to prevent memory issues during extended offline
//   - All outbound messages logged through WatchLogger with source "SignalR"
//
// Outbound messages:
//   - "TestStepCompleted" → test step result back to orchestrator
//   - "LogEntry" → streams log entry to dashboard
//   - "DeviceStatusUpdate" → periodic device health report
//   - "Pong" → response to server Ping
//   - "ReportAdapterTiers" → current adapter tier map
//   - "UpdateResponderLocation" → en-route location updates
//   - "ResponderOnScene" → arrival notification
//   - "NotifyEvidenceSubmitted" → evidence upload notification
//
// Example:
// ```swift
// let sender = HubEventSender.shared
//
// // Send a test step result
// sender.sendTestStepCompleted(
//     runId: "run_abc123",
//     stepId: "step_001",
//     passed: true,
//     durationMs: 1250
// )
//
// // Stream a log entry
// sender.sendLogEntry(
//     level: "Information",
//     source: "LocationCoordinator",
//     message: "Mode escalated to Emergency",
//     correlationId: "sos_xyz"
// )
//
// // Messages queue if disconnected and auto-flush on reconnect
// ```

/// Queued message waiting to be sent when connection is restored.
struct QueuedHubMessage: Sendable {
    let method: String
    let argsJson: String
    let queuedAt: Date
}

/// Outbound message sender for the SignalR hub.
///
/// Provides typed methods for every message the mobile client sends to the
/// DashboardHub. Handles offline queueing and reconnect flushing automatically.
///
/// Thread safety: all sends are serialized through a DispatchQueue to prevent
/// concurrent writes on the SignalR connection.
final class HubEventSender: @unchecked Sendable, HubConnectionDelegate {

    static let shared = HubEventSender()

    private let hub = WatchHubConnection.shared
    private let logger = WatchLogger.shared
    private let sendQueue = DispatchQueue(label: "com.thewatch.signalr.sender", qos: .userInitiated)

    // ── Offline Queue ────────────────────────────────────────────

    private var messageQueue: [QueuedHubMessage] = []
    private let maxQueueSize = 500

    private let iso8601: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private init() {
        hub.addDelegate(self)
    }

    // ── HubConnectionDelegate ────────────────────────────────────

    func hubDidConnect() {
        sendQueue.async { [weak self] in
            self?.flushQueue()
        }
    }

    func hubDidDisconnect(error: Error?) {}
    func hubIsReconnecting(attempt: Int) {}

    // ── Test Orchestration ───────────────────────────────────────

    /// Send a test step result back to the TestOrchestratorService.
    func sendTestStepCompleted(
        runId: String,
        stepId: String,
        passed: Bool,
        screenshot: String? = nil,
        errorMessage: String? = nil,
        durationMs: Int64 = 0
    ) {
        var args: [String: String] = [
            "runId": runId,
            "stepId": stepId,
            "passed": "\(passed)",
            "durationMs": "\(durationMs)"
        ]
        if let screenshot { args["screenshot"] = screenshot }
        if let errorMessage { args["errorMessage"] = errorMessage }

        send(method: "TestStepCompleted", args: args, correlationId: runId)
    }

    // ── Log Streaming ────────────────────────────────────────────

    /// Stream a log entry to the dashboard for real-time log viewing.
    func sendLogEntry(
        level: String,
        source: String,
        message: String,
        properties: [String: String] = [:],
        exception: String? = nil,
        correlationId: String? = nil
    ) {
        var args: [String: String] = [
            "level": level,
            "source": source,
            "message": message,
            "timestamp": iso8601.string(from: Date())
        ]
        if !properties.isEmpty, let data = try? JSONEncoder().encode(properties),
           let json = String(data: data, encoding: .utf8) {
            args["properties"] = json
        }
        if let exception { args["exception"] = exception }
        if let correlationId { args["correlationId"] = correlationId }

        send(method: "LogEntry", args: args, correlationId: correlationId)
    }

    // ── Device Status ────────────────────────────────────────────

    /// Send a periodic device status update to the dashboard.
    func sendDeviceStatusUpdate(
        batteryPercent: Int,
        networkType: String,
        gpsEnabled: Bool,
        adapterTiers: [String: String],
        freeMemoryMb: Int64
    ) {
        var args: [String: String] = [
            "batteryPercent": "\(batteryPercent)",
            "networkType": networkType,
            "gpsEnabled": "\(gpsEnabled)",
            "freeMemoryMb": "\(freeMemoryMb)",
            "timestamp": iso8601.string(from: Date())
        ]
        if let data = try? JSONEncoder().encode(adapterTiers),
           let json = String(data: data, encoding: .utf8) {
            args["adapterTiers"] = json
        }

        send(method: "DeviceStatusUpdate", args: args)
    }

    // ── Keepalive ────────────────────────────────────────────────

    /// Respond to a server Ping with Pong. Not queued.
    func sendPong() {
        guard hub.connectionState.isConnected else { return }

        sendQueue.async { [weak self] in
            // (self?.hub.hubConnection as? SignalRClient.HubConnection)?.invoke(method: "Pong", Date().iso8601)
            self?.logger.verbose(source: "SignalR", template: "Sent Pong")
        }
    }

    // ── Adapter Tier Reporting ────────────────────────────────────

    /// Report current adapter tier configuration to the dashboard.
    func sendReportAdapterTiers(_ tiers: [String: String]) {
        if let data = try? JSONEncoder().encode(tiers),
           let json = String(data: data, encoding: .utf8) {
            send(method: "ReportAdapterTiers", args: ["tiers": json])
        }
    }

    // ── Response Coordination ────────────────────────────────────

    /// Send a responder location update while en-route.
    func sendResponderLocation(
        requestId: String,
        responderId: String,
        latitude: Double,
        longitude: Double,
        speedMps: Double? = nil
    ) {
        var args: [String: String] = [
            "requestId": requestId,
            "responderId": responderId,
            "latitude": "\(latitude)",
            "longitude": "\(longitude)",
            "timestamp": iso8601.string(from: Date())
        ]
        if let speedMps { args["speedMps"] = "\(speedMps)" }

        send(method: "UpdateResponderLocation", args: args, correlationId: requestId)
    }

    /// Signal that the responder has arrived on scene.
    func sendResponderOnScene(requestId: String, responderId: String) {
        let args: [String: String] = [
            "requestId": requestId,
            "responderId": responderId,
            "arrivedAt": iso8601.string(from: Date())
        ]
        send(method: "ResponderOnScene", args: args, correlationId: requestId)
    }

    /// Notify the dashboard that evidence has been submitted.
    func sendEvidenceSubmitted(
        submissionId: String,
        requestId: String,
        phase: String,
        type: String,
        thumbnailUrl: String? = nil
    ) {
        var args: [String: String] = [
            "submissionId": submissionId,
            "requestId": requestId,
            "phase": phase,
            "type": type,
            "timestamp": iso8601.string(from: Date())
        ]
        if let thumbnailUrl { args["thumbnailUrl"] = thumbnailUrl }

        send(method: "NotifyEvidenceSubmitted", args: args, correlationId: requestId)
    }

    // ── Core Send / Queue ────────────────────────────────────────

    private func send(method: String, args: [String: String], correlationId: String? = nil) {
        guard hub.connectionState.isConnected else {
            enqueue(method: method, args: args)
            return
        }

        sendQueue.async { [weak self] in
            guard let self else { return }

            // let connection = self.hub.hubConnection as? SignalRClient.HubConnection
            // let argsJson = self.encodeArgs(args)
            // connection?.invoke(method: method, argsJson) { error in
            //     if let error {
            //         self.logger.warning(...)
            //         self.enqueue(method: method, args: args)
            //     }
            // }

            let argsJson = self.encodeArgs(args)
            self.logger.debug(
                source: "SignalR",
                template: "Sent {Method} ({Bytes} bytes)",
                properties: [
                    "Method": method,
                    "Bytes": "\(argsJson.count)"
                ],
                correlationId: correlationId
            )
        }
    }

    private func enqueue(method: String, args: [String: String]) {
        let argsJson = encodeArgs(args)
        let message = QueuedHubMessage(method: method, argsJson: argsJson, queuedAt: Date())

        sendQueue.async { [weak self] in
            guard let self else { return }

            self.messageQueue.append(message)

            // Trim if over capacity (drop oldest)
            while self.messageQueue.count > self.maxQueueSize {
                let dropped = self.messageQueue.removeFirst()
                self.logger.warning(
                    source: "SignalR",
                    template: "Queue overflow: dropped {Method} queued at {QueuedAt}",
                    properties: [
                        "Method": dropped.method,
                        "QueuedAt": self.iso8601.string(from: dropped.queuedAt)
                    ]
                )
            }

            self.logger.debug(
                source: "SignalR",
                template: "Queued {Method} for delivery on reconnect. Queue size: {Size}",
                properties: [
                    "Method": method,
                    "Size": "\(self.messageQueue.count)"
                ]
            )
        }
    }

    /// Flush all queued messages after reconnection. Messages sent FIFO.
    private func flushQueue() {
        guard !messageQueue.isEmpty else { return }

        let count = messageQueue.count
        logger.information(
            source: "SignalR",
            template: "Flushing {Count} queued messages after reconnect",
            properties: ["Count": "\(count)"]
        )

        var flushed = 0
        var failed = 0

        while !messageQueue.isEmpty {
            let message = messageQueue.removeFirst()

            // let connection = hub.hubConnection as? SignalRClient.HubConnection
            // connection?.invoke(method: message.method, message.argsJson) { error in ... }
            flushed += 1
        }

        logger.information(
            source: "SignalR",
            template: "Queue flush complete: {Flushed} sent, {Failed} failed",
            properties: [
                "Flushed": "\(flushed)",
                "Failed": "\(failed)"
            ]
        )
    }

    // ── Utility ──────────────────────────────────────────────────

    private func encodeArgs(_ args: [String: String]) -> String {
        (try? String(data: JSONEncoder().encode(args), encoding: .utf8)) ?? "{}"
    }

    /// Returns queue diagnostics.
    func diagnosticsSnapshot() -> [String: String] {
        [
            "QueueSize": "\(messageQueue.count)",
            "MaxQueueSize": "\(maxQueueSize)",
            "ConnectionState": hub.connectionState.rawValue
        ]
    }
}
